src/approaches/grammar_search_invention_approach.py

```diff
@@ -828,8 +852,24 @@ def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
             # Bad case 2: transition is not optimal and sequence is a demo.
             elif on_sequence_h > best_h and is_demo:
                 score += CFG.grammar_search_on_demo_count_penalty
+            # Bad case 3: there is a "suspicious" optimal state.
+            for successor in optimal_successors:
+                # We already know that next_atoms is not suspicious.
+                if successor == frozenset(next_atoms):
+                    continue
+                if self._is_suspicious(successor, demo_atom_sets):
+                    score += CFG.grammar_search_suspicious_penalty
         return score
 
+    @staticmethod
+    def _is_suspicious(successor: FrozenSet[GroundAtom],
```

Comment:
i guess one way of thinking about this is that we're asserting that the set of lifted states seen in the demonstrations is the set of "good" states, and any candidate predicate set which believes that a state outside this set is also good would be a bad predicate set.

Reply:
right exactly